momentGW/pbc/tda.py

```diff
@@ -66,23 +92,42 @@ def build_dd_moments(self):
                         (self.mo_occ_w[kj], self.mo_occ_w[kb]),
                     )
                     moments[q, kb, i] += moments[q, kb, i - 1] * d.ravel()[None]
+                    if self.fc and q == 0:
+                        head[kj, i] += head[kj, i - 1] * d.ravel()
 
                 tmp = np.zeros((self.naux[q], self.naux[q]), dtype=complex)
+                tmp_head = np.zeros((self.naux[q]), dtype=complex)
+
                 for ki in kpts.loop(1, mpi=True):
                     ka = kpts.member(kpts.wrap_around(kpts[q] + kpts[ki]))
 
                     tmp += np.dot(moments[q, ka, i - 1], self.integrals.Lia[ki, ka].T.conj())
 
+                    if q == 0 and self.fc:
+                        tmp_head += util.einsum(
+                            "a,aP->P", head[ki, i - 1], self.integrals.Lia[ki, ki].T.conj()
+                        )
+
                 tmp = mpi_helper.allreduce(tmp)
                 tmp *= 2.0
                 tmp /= self.nkpts
 
+                tmp_head = mpi_helper.allreduce(tmp_head)
+                tmp_head *= 2.0
+                tmp_head /= self.nkpts
+
                 for kj in kpts.loop(1, mpi=True):
                     kb = kpts.member(kpts.wrap_around(kpts[q] + kpts[kj]))
 
                     moments[q, kb, i] += np.dot(tmp, self.integrals.Lai[kj, kb].conj())
 
-        return moments
+                    if q == 0 and self.fc:
+                        head[kj, i] += util.einsum("P,Pa->a", tmp_head, self.integrals.Lia[kj, kj])
+
+        if self.fc:
+            return {"moments": moments, "head": head}
+        else:
+            return moments
```

Comment:
i'm not a huge fan of how this function looks now -- it's become too convoluted, and having multiple return types depending on the `fc` setting is not good practice. Is it possible to just have a `get_moments_dd_head` function that does the head part? Will that be significantly more inefficient?